### PR TITLE
Embed files in RFML - ETA 8/5

### LINKED ISSUE #161: Embed files in RFML
In the web app you can drag and drop files into tests to have them embedded (as screenshots or files). We should be able to embed files easily with RFML as well.

### PR DESCRIPTION
**Pull request by epaulet**
fixes #161

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -179,6 +182,14 @@ tests and the first step of a test.
 - `EMBEDDED TEST RFML ID` - Embed the steps of another test within the current test
 using the embedded test's RFML ID.
 
+Embedding Screenshots and Downloadable Files:
+- `{{ file.screenshot(RELATIVE_PATH) }}` will embed screenshot found at RELATIVE_PATH
+which is a path to the screenshot relative to the RFML file.
+- `{{ file.download(RELATIVE_PATH) }}` will embed downloadable file link found at RELATIVE_PATH
+which is a path to the downloadable file relative to the RFML file.
+- It is _strongly_ recommended that your embedded files are checked into source control with the rest
```
Comment (ukd1) [inline README.md]: isn't it required? what happens if the file isn't represent?

Comment (epaulet) [inline README.md]: Probably should be. Otherwise everyone need to have a copy of it. Will update.

File: lib/rainforest_cli/test_parser/step.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+class RainforestCli::TestParser::Step < Struct.new(:action, :response, :redirect)
+  UPLOADABLE_REGEX = /{{ *file\.(download|screenshot)\(([^\)]+)\) *}}/
```
Comment (ukd1) [inline lib/rainforest_cli/test_parser/step.rb]: totally just randomly thought; is ``file`` disallowed for custom variables?

File: lib/rainforest_cli/uploader.rb [hunk 3]
```diff
@@ -26,6 +28,30 @@ def upload
...
   end
 
+  def create_test_obj(rfml_test)
+    parse_uploadables!(rfml_test) if rfml_test.has_uploadable_files?
+
+    elements = rfml_test.steps.map do |step|
+      element = case step.type
+                when :test then { id: primary_key_dictionary[step.rfml_id] }
+                when :step then { action: step.action, response: step.response }
+                end
```
Comment (ukd1) [inline lib/rainforest_cli/uploader.rb]: can we have a default that crashes?

File: lib/rainforest_cli/uploader/multi_form_post_request.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+require 'httparty'
+
+class RainforestCli::Uploader::MultiFormPostRequest
```
Comment (ukd1) [inline lib/rainforest_cli/uploader/multi_form_post_request.rb]: uggh this is so annoying you had to do all this work; looks great though!

File: lib/rainforest_cli/uploader/uploadable_parser.rb [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+      'key' => aws_info['aws_key'],
+      'AWSAccessKeyId' => aws_info['aws_access_id'],
+      'acl' => aws_info['aws_acl'],
+      'policy' => aws_info['aws_policy'],
+      'signature' => aws_info['aws_signature'],
+      'Content-Type' => MimeMagic.by_path(file),
+      'file' => file,
+    )
+
+    unless resp.code.between?(200, 299)
```
Comment (ukd1) [inline lib/rainforest_cli/uploader/uploadable_parser.rb]: isn't between ``x < y`` i.e. 299 should be 300?

**Comment by ukd1** [comment]
lgtm to me, but I don't see any negative tests - what happens when the local file doesn't exist?

File: lib/rainforest_cli/test_parser/step.rb (hunk 2 above)
Comment (epaulet) [inline lib/rainforest_cli/test_parser/step.rb]: The API shouldn't it probably does currently. Will check separately.

File: lib/rainforest_cli/uploader/multi_form_post_request.rb (hunk 4 above)
Comment (epaulet) [inline lib/rainforest_cli/uploader/multi_form_post_request.rb]: I'll circle back later and replace with an existing solution.

File: lib/rainforest_cli/uploader/uploadable_parser.rb (hunk 5 above)
Comment (epaulet) [inline lib/rainforest_cli/uploader/uploadable_parser.rb]: It's inclusive `200 <= x <= 299`

**Comment by epaulet** [comment]
It doesn't replace the text at all: https://github.com/rainforestapp/rainforest-cli/pull/163/files#diff-f287bd36c3fd8a24f4ad0bd7616b9199R36

**Comment by epaulet** [comment]
Okay, added the negative tests - merging now 💃

**Pull request merged**